### PR TITLE
Support Multi-Variate Regression

### PR DESCRIPTION
**Pull request by nilsleh**
This PR aims to extend Regression support to Multivariate Regression Cases. Mostly this involves some "tensor gymnastics" because there is now an additional dimension to be handled and we try to cover possible exemplary outputs

- batch_size X 1 (mse output single taret)
- batch_size X 2 (nll output single target or multivariate output for mse loss)
- batch_size X 5 (mse multivariate with 5 targets)
- batch_size X 2 X 5 (nll output for each of 5 targets)
- batch_size X 2 X 5 X 7 (nll output for 5 targets with 7 samples like in MC-Dropout or Ensembles)

Current breaking points that this PR has to address:
- automatic logging of multi-variate targets. We largely rely on torchmetrics for default metric computation. They have a [MultiOutput Wrapper](https://lightning.ai/docs/torchmetrics/stable/wrappers/multi_output_wrapper.html) but still have to decide between logging a mean over the different outputs per metric (logging would still show a single metric but one wouldn't know how individual targets are doing) or logging each multivariate output as its own logging entry (train_RMSE_1, train_RMSE_2,...) this would yield a lot of logging and plots potentially so cluttered
- processing of regression predictions like examples given above, there are some special cases that make life difficult like knowing when to do handle NLL predictions (however, this might involve some code duplications to clearly separate the cases)
- ...
- some methods like Laplace will also have different outputs that need to be handled

Overall, I think N-D regression can be integrated into the existing 1D Regression classes *MethodName*Regression by adding an additional `n_targets` argument that specifies the number of regression targets and can be utilized to handle the different output cases. It's certainly a good amount of work, but I think the unittesting is good enough, to make such changes and make sure everything else implemented already does not break. It would certainly extend the number of possible applications and hopefully users :) 

Very much open to suggestions and implementation ideas.

### CODE REVIEW

**Comment by nilsleh** [comment]
@codingS3b I got curious how one would go about N-D regression and to what extend it's possible with the current state of the library. So as you said there are errors both when using the `Regression` or `PxRegression` task, because the logged metrics are different in those cases, as outlined above. With the changes included in this PR branch, I can now run a Toy example:

```python
from functools import partial
import torch
from lightning import Trainer
from lightning_uq_box.models import MLP
from lightning_uq_box.uq_methods import MVERegression
from lightning_uq_box.datamodules import ToyMultiRegressionDataModule

if __name__ == "__main__":
    num_features = 10
    num_targets = 5
    datamodule = ToyMultiRegressionDataModule(
        n_features=num_features, n_targets=num_targets, batch_size=32
    )

    # n_ouputs is 2 because we are predicting mean and variance
    model = MLP(n_inputs=num_features, n_outputs=2, n_targets=num_targets)

    mve = MVERegression(
        model=model,
        optimizer=partial(torch.optim.Adam, lr=1e-4),
        burnin_epochs=0,
        n_targets=5,
    )
    trainer = Trainer(
        max_epochs=10,
        accelerator="cpu",
        logger=True,
        log_every_n_steps=1,
        default_root_dir="logs",
    )

    trainer.fit(mve, datamodule)

    trainer.test(mve, datamodule)
```

I haven't worked on N-D regression problems so not sure what else is required but I'd be open to work on supporting the task for the other methods as well. Let me know what you think and how we can help.

@elcorto also tagging you since you mentioned that you are also looking into N-D regression for a project.

**Comment by codingS3b** [comment]
Thanks for looking into this, @nilsleh! I think, for the use case we recently discussed (input shape: B x 1 x W, output shape: B x 1 x W),  I managed to stitch together a deep ensemble using the tools provided in the main branch, which then did work out of the box :) 

Specifically, I did the following:
- use `MVEPxRegression` in combination with a 1D-CNN model that keeps the shape of the input intact and produces two output channels (one for mean, one for sigma)
- use `DeepEnsemblePxRegression` with multiple trained instances of the above

However, when there is a restriction to the use of MLP`s, your outlined approach might indeed be needed, but I'm not sure.

**Comment by codingS3b** [comment]
@nilsleh, if you want I could contribute the notebook I created as an additional tutorial? But I guess it will need a bit of polishing then...Let me know if you are interested.

**Comment by nilsleh** [comment]
I think a tutorial notebook would be great. But also think it would be cleaner to rework the library a bit to support multi-variate regression inside the Regression module and leave the `PxRegression` module as is. But this would require a bit of work, and not sure when I can get to it (probably after next week).

**Comment by codingS3b** [comment]
I'm struggling to see the differences (or pros and cons) between both approaches, maybe you could elaborate a bit:
a) tackling a multivariate regression problem with a 1D-CNN and the `PxRegression` (which already works and I'm still owing you the promised notebook :) )
and
b) using (a potentially enhanced version of) `MVERegression` (with an MLP?) which would require, as you stated above, a code refactoring  

I'm already happy that a) works and do not really see what would be different if you implemented b). Greatly appreciate your thoughts on this!

**Comment by nilsleh** [comment]
@codingS3b Thanks for your inpup. I will try to elaborate a bit. 

1. When first thinking about how to implement various methods for different tasks (regression, classification, segmentation), it became clear that some things are shared across tasks and methods and we wanted to support those out of the box. The main difference between these different tasks is the output/target modality data, which can be 1D like single regression or classification, or 2D for multivariate regression or 3D for semantic segmentation. However, each of them usually have some commonalities for which we want to provide "out-of-the-box" support with respect to training/evalution. For example, regardless of the UQ method, a segmentation task will require a certain set of accuracy metrics where we provide some [defaults](https://github.com/lightning-uq-box/lightning-uq-box/blob/8cdcbd3e6441ab50cf664dc6f65878f36338be75/lightning_uq_box/uq_methods/base.py#L422) so that people can just start training. This would also apply to logging samples during training (2D vs 3D).  Similarly, we wanted to provide some defaults to save results for post analysis like [here](https://github.com/lightning-uq-box/lightning-uq-box/blob/8cdcbd3e6441ab50cf664dc6f65878f36338be75/lightning_uq_box/uq_methods/base.py#L466) and this later part is dependent on the data modality (vector vs image), where for vector data we chose csv and for image data hdf5. So to be more concrete `PxRegression` was "designed" or intented to work with 3D image data, and while you were able to make it work, it did require some hacks I presume and I would favor a more "out-of-the-box" solution that is also more clear with respect to naming.
2. There is additional functionality for the 3D tasks like `PxRegression` and `SemanticSegmentation` like the `freeze` arguments which is a bit specific to the types of models commonly used for this target data. With everything we tried not to reinvent the wheel so for those 3D tasks, strong or even SOTA models are usually some UNet as available in [torchseg](https://github.com/isaaccorley/torchseg) for which it can be very convenient to freeze a part of the model (encoder or decoder) for finetuning while allowing for UQ. So these are just some additional functionalities specific for 3D outputs and are not default applicable for 2D multivariate targets. We tried to implement everything such that it can be easily overwritten if one ones to change one of these aspects, however, it seems "not well designed" if users are required to always overwrite functionality to just run an experiment. 
3. From the Earth Observation Side, Semantic Segmentation and Pixelwise Regression are two central tasks and there are actually not that many methods that work with such image data (in contrast to single or multivariate vector regression, where more methods exists). Therefore, we wanted to clearly emphasize available methods for those input/output modalities. If `PxRegression` should include multivariate regression, edge cases arise. A concrete example would be the `LaplaceApproximation` which works for vector data, but not image output targets (it could in principle but computationally to demanding). So for `Laplace` we would like to support multivariate regression, but naming it `PxRegression` would feel imprecise.

So to summarize, I would like to follow a clear naming structure of classes that makes intuitive sense and align with the "task-specific" and target data type with the additional functionality that we thought of. Because of that having Multivariate regression under `Regression` seems more appropriate to me than `PxRegression` (2D vs 3D data).

With respect to your question, the `Regression` modules don't have any restriction to the model architecture, generally, we tried to be "pytorch model agnostic" in the sense that we "don't care" what the input modality is, as long as your pytorch model can transform that input to the required target shape. So `Regression` is not constrained to `MLPs`. For example, one can have a BNN with a pretrained Resnet18 (CNN model) like in this [test config](https://github.com/lightning-uq-box/lightning-uq-box/blob/main/tests/configs/image_regression/bnn_vi_elbo.yaml). The naming of `Regression`, `PxRegression`, `Classification` etc is not so much about the type of input data, but the modality of the target data (1D, 2D, 3D). 

I am hoping that this is not too confusing with all this input, and I am also not claiming that this is the best way of implementing everything, so very much up for discussing this and I realize there is much room for improvement for our documentation. Maybe this elaboration provides a better playing field for a discussion. :)
